Support to specify Strings that are hashed as Inputs via a commandline parameter

StephenDiligent opened the pull request:
**Disclaimer:** this is the first GoLang I have ever written so please review in detail and rest assured that all feedback is welcomed.

These changes address issue https://github.com/simplesurance/baur/issues/119
For now I have used the `additional-input-str` and `lookup-additional-input-str-fallback` parameter names outlined in the issue. However, I am more than happy to changes these parameter names as you see fit.
To reduce complexity I have only added support for a single `lookup-additional-input-str-fallback` value instead of accepting values in the format `VAL[,VAL]...`.
The `run`, `status` and `ls inputs` commands have been updated to support these new parameters.

Example outputs:
![image](https://user-images.githubusercontent.com/20102859/92349569-a1946980-f12a-11ea-9667-b7cc86918024.png)
![image](https://user-images.githubusercontent.com/20102859/92349600-bb35b100-f12a-11ea-8cf1-f10077414d5a.png)
![image](https://user-images.githubusercontent.com/20102859/92349643-da344300-f12a-11ea-8e59-2e8992501c2f.png)
![image](https://user-images.githubusercontent.com/20102859/92349710-fd5ef280-f12a-11ea-913d-a60f1ef55b0c.png)
![image](https://user-images.githubusercontent.com/20102859/92349727-094ab480-f12b-11ea-85e6-8282c82c0fc9.png)

The additional input strings are stored as `additional_string: <value>` in the DB
![image](https://user-images.githubusercontent.com/20102859/92349789-38612600-f12b-11ea-9014-30d1566048ac.png)

The `setting version failed` message does not appear when Baur is compiled into a binary.
Also, you may have noticed in the screenshots that I am using PowerShell on Windows. I plan on raising a PR to support Windows in the future if you would be happy to accept it.

fho commented:
Hello @StephenDiligent,

thanks a lot for your PR. I'll review it soon.

Yes, I'm interested in supporting Windows.
To ensure it does not break we should also modify the CI setup to run all the testcases on a Windows VM.

fho reviewed (changes requested):
Good job, the code looks good, especially for being your first Go implementation!

The current implementation couples the Inputs and Resolver objects with the storage.
Both should not be coupled to the storage and only depend on the state of the app/task in the current repository.

I suggest to implement it in the following way instead:
- Resolver: keep it as it was. The responsibility of the InputResolver is to resolve abstract input definitions like Glob patterns to specific inputs. There is nothing to resolve for `AdditionalStr`.
- Inputs: add a method like `func (in *Inputs) AddStringInput(str string)` to add `AdditionalStr` as an Input
- TaskStatusResolver: 
  - call inputResolver.Resolve()
  - call inputs.AddStringInput(`AdditionalStr`), check if a record for it's total digest exit in the db via, `t.store.LatestTaskRunByDigest()`,
  - if not add `lookup-additional-input-str-fallback` to the Inputs object instead and check if a result for that totaldigest exists in the db,
- always return the Inputs object that contains `AdditionalStr` from `*TaskStatusEvaluator.Status()`

The point of adding `AdditionalStr` to the configuration files (`cfg` package) is not clear to me.
Afaik the idea is to only pass those additional input string via the commandline. What is the usecase for also specifying them in the configuration file?

>  For now I have used the additional-input-str and lookup-additional-input-str-fallback parameter names outlined in the issue. However, I am more than happy to changes these parameter names as you see fit.

I do not have good parameter names in mind right now, we definitely need shorter ones though. :-)
We can change the parameter names and other details later.

>  To reduce complexity I have only added support for a single lookup-additional-input-str-fallback value instead of accepting values in the format VAL[,VAL]....

ok, fine with me :+1: 

> The setting version failed message does not appear when Baur is compiled into a binary.

If you build the binary via the Makefile (`make`) the version will be set.
The version is set via ldflags during build (https://github.com/simplesurance/baur/blob/master/Makefile#L4)

StephenDiligent commented:
> The current implementation couples the Inputs and Resolver objects with the storage.
> Both should not be coupled to the storage and only depend on the state of the app/task in the current repository.

I agree, it did feel dirty at the time pushing the store down into `Inputs` and `InputResolver`. 
I have updated the implementation as per your suggestion.

> The point of adding `AdditionalStr` to the configuration files (`cfg` package) is not clear to me.
> Afaik the idea is to only pass those additional input string via the commandline. What is the usecase for also specifying them in the configuration file?

It is not required, this was due to my unfamiliarity with the codebase. I have removed these changes.

> I do not have good parameter names in mind right now, we definitely need shorter ones though. :-)
> We can change the parameter names and other details later.

All of the names that I have come up with are just as long as `AdditionalInputStr` e.g. `AncillaryInputStr`, `SupplementaryInputStr`. 
Perhaps use `--add-input-str=feature-x` and `--alt-input-str-lookup=master` (alt == alternative) 🤷

StephenDiligent commented:
@fho Have you had a chance to look over the latest changes?

fho commented:
@StephenDiligent sorry for the late response, I was on sick leave the last days. I'll have a look soon.

fho commented on internal/command/run.go:
```diff
@@ -95,6 +97,10 @@ func newRunCmd() *runCmd {
+	cmd.Flags().StringVar(&cmd.additionalInputStr, "additional-input-str", "",
+		"include an additional string as an input")
+	cmd.Flags().StringVar(&cmd.lookupAdditionalInputStrFallback, "lookup-additional-input-str-fallback", "",
+		"include an additional input string to fallback to if a run is not found with the additional-input-str value provided")
```
```suggestion
		"if a run can not be found, try to find a run with this value as input-string")
```

fho reviewed (commented):
great improvements, I also did some manual tests and it works as expected
I added some comments about minor issues.

> All of the names that I have come up with are just as long as AdditionalInputStr e.g. AncillaryInputStr, SupplementaryInputStr.
Perhaps use --add-input-str=feature-x and --alt-input-str-lookup=master (alt == alternative) shrug

:+1: 
Let's go with `--input-str=feature-x` and `--alt-input-str` and also name the fields in the structs accordingly.

Could you also add a testcase that calls the `baur run` and `baur status` commands with the new parameters and checks if the the behavior is expected?
You can find examples `internal/command/command_test.go`.
Alternatively, if you grant me access to your fork branch I could also add them.
(https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)

StephenDiligent commented on inputs.go:
```diff
@@ -9,11 +9,23 @@ import (
-	Files  []*Inputfile
-	digest *digest.Digest
+	Files         []*Inputfile
+	AdditionalStr *InputString
```
No problem, I'll make this change. 
I also want to point out that the `Files` field is also "public" and if this field is updated after the digest has been calculated the value returned from the `Digest()` function will not represent the updated `Files` field value.

fho replied:
Right, same issue, we should also make it private

StephenDiligent commented on input.go:
```diff
@@ -0,0 +1,8 @@
+import "github.com/simplesurance/baur/v1/internal/digest"
+
+// Input represents an input
+type Input interface {
```
It is also used in `taskstatusevaluator.go` so it cannot be moved down into `internal/command/ls_inputs.go`

StephenDiligent commented on inputs.go:
```diff
@@ -9,11 +9,23 @@ import (
 
-// Digest returns a summarized digest over all Files.
+// NewInputs returns a new Inputs
+func NewInputs(files []*Inputfile) *Inputs {
```
I assume that an empty `Inputs` with no files is not a valid state and a "constructor" should ideally be used. I have replaced the instance of `&Inputs` being instantiated in `inputresolver.go` with this constructor.
I was following this guide https://golang.org/doc/effective_go.html#composite_literals
I'm happy to revert this change if you want me to.

fho replied:
:+1: ok, makes sense, lets keep it :-)

StephenDiligent commented on inputs.go:
```diff
@@ -9,11 +9,23 @@ import (
+}
+
+// AddAdditionalString adds a string value to AdditionalStr
+func (in *Inputs) AddAdditionalString(additionalStr string) {
```
Done

StephenDiligent commented on taskstatusevaluator.go:
```diff
@@ -43,19 +50,36 @@ func (t *TaskStatusEvaluator) Status(ctx context.Context, task *Task) (TaskStatu
+	inputs.AddAdditionalString(t.additionalInputStr)
+	taskStatus, run, err = t.getTaskStatus(ctx, inputs, task)
+
+	if run == nil && t.lookupAdditionalInputStrFallback != "" {
```
Done

StephenDiligent commented:
> Could you also add a testcase that calls the `baur run` and `baur status` commands with the new parameters and checks if the the behavior is expected?
> You can find examples `internal/command/command_test.go`.
> Alternatively, if you grant me access to your fork branch I could also add them.
> (https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)

I'll try and add these tests now. 
You should be able to add commits to my branch as I have the _Allow edits by maintainers
_ option checked.
![image](https://user-images.githubusercontent.com/20102859/93414887-5f231780-f8f6-11ea-98ef-6447963fb57c.png)

fho commented on inputs.go:
```diff
@@ -31,6 +48,15 @@ func (in *Inputs) Digest() (*digest.Digest, error) {
 		digests[i] = fdigest
 	}
 
+	if in.inputString.Exists() {
```
```suggestion
	if in.inputString != nil && in.inputString.Exists() {
```
The constructor does not initialize inputString. If `Digest()` is called afterwards without calling `SetInputString()` a segfault would happen.

StephenDiligent replied:
This has been resolved in b756e3b

StephenDiligent commented:
@fho I have added the requested tests in 5d8a4e4

fho reviewed (approved):
Thanks a lot again for the implementation and the tests! :tada: 

I tried to push some changes to your branches but could not because I did not have permissions (or I did something wrong?).
I'll merge the PR as it is and push the changes in a followup.
